### PR TITLE
Use `Atomic` database by default in `clickhouse-local`

### LINKED ISSUE #44817: CREATE OR REPLACE and EXCHANGE TABLES for Memory database
**Use case**

`Memory` database engine is used in `clickhouse-local`, and nothing prevents implementing these queries.

**Describe the solution you'd like**

A mutex.

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use `Atomic` database by default in `clickhouse-local`. Address items 1 and 5 from #50647. Closes #44817.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 02928bb207348019a0968454e0f7d6baa6e511af with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65860/02928bb207348019a0968454e0f7d6baa6e511af/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/65120/0dd4735e07319295cf52043bf9fd1fd71f0fc1ea/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by jsc0218** [comment]
looks like the [report](https://d1k2gkhrlfqv31.cloudfront.net/clickhouse-test-reports-private/11845/17c6b97cbcb90e19c236708116f20ef3f88cc9c1/integration_tests__asan__old_analyzer__[4_6].html) from private CI can manifest the potential issue.

File: tests/queries/0_stateless/02141_clickhouse_local_interactive_table.reference [hunk 1]
```diff
@@ -1,2 +1,2 @@
-CREATE TABLE default.`table`\n(\n    `key` String\n)\nENGINE = File(\'TSVWithNamesAndTypes\', \'/dev/null\')
-CREATE TABLE foo.`table`\n(\n    `key` String\n)\nENGINE = File(\'TSVWithNamesAndTypes\', \'/dev/null\')
```
Comment (vitlibar) [inline tests/queries/0_stateless/02141_clickhouse_local_interactive_table.reference:2]: What was wrong with non-temporary table here?

File: src/Databases/DatabaseAtomic.cpp [hunk 2]
```diff
@@ -53,16 +53,23 @@ DatabaseAtomic::DatabaseAtomic(String name_, String metadata_path_, UUID uuid, c
     , db_uuid(uuid)
 {
     assert(db_uuid != UUIDHelpers::Nil);
-    fs::create_directories(fs::path(getContext()->getPath()) / "metadata");
-    fs::create_directories(path_to_table_symlinks);
-    tryCreateMetadataSymlink();
```
Comment (vitlibar) [inline src/Databases/DatabaseAtomic.cpp:58]: Why can't we create directories in the constructor now?

File: tests/queries/0_stateless/02141_clickhouse_local_interactive_table.reference (hunk 1 above)
Comment (alexey-milovidov) [reply tests/queries/0_stateless/02141_clickhouse_local_interactive_table.reference:2]: Now clickhouse-local uses temporary tables for inputs.

File: src/Databases/DatabaseAtomic.cpp (hunk 2 above)
Comment (alexey-milovidov) [reply src/Databases/DatabaseAtomic.cpp:58]: clickhouse-local is used most of the time without using the database, and creating and removing a lot of directories slows it down.

File: src/Databases/DatabasesOverlay.cpp [hunk 3]
```diff
@@ -178,6 +213,18 @@ String DatabasesOverlay::getTableDataPath(const ASTCreateQuery & query) const
...
+UUID DatabasesOverlay::getUUID() const
+{
+    UUID result = UUIDHelpers::Nil;
+    for (const auto & db : databases)
+    {
+        result = db->getUUID();
+        if (result != UUIDHelpers::Nil)
+            break;
+    }
+    return result;
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Databases/DatabasesOverlay.cpp:225]: The question is, what is this UUID for? - It is needed for Atomic and Replicated databases to allow database rename and to automatically choose paths in Keeper for Replicated tables. When there is an overlay of Atomic/Replicated and one ephemeral database, using the same UUID looks reasonable.

**Comment by alexey-milovidov** [comment]
FYI, for the race condition of settings in Context in clickhouse-local, this pull request is relevant: https://github.com/ClickHouse/ClickHouse/pull/60197

**Comment by vitlibar** [comment]
> FYI, for the race condition of settings in Context in clickhouse-local, this pull request is relevant: #60197

I've already fixed the race condition in https://github.com/ClickHouse/ClickHouse/pull/67133

File: src/Databases/DatabasesOverlay.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/Databases/DatabasesOverlay.cpp:225]: That looks quite random.

Also when a database is attached its UUID is removed from the global map. Due to now we can have two databases with the same UUID: `Overlay` and `Atomic/Replicated`, will it try to remove the same UUID twice?

Maybe it's better to simply give `Overlay` a random `UUID`?

[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Databases/DatabasesOverlay.cpp:225]: That's true - sounds dangerous. However, it is not possible to create an Overlay database now, and it is available only in clickhouse-local.

I will check if the tests in this PR will be green. I will merge it with this remaining consideration and will try to fix it in a follow-up PR.

**Comment by azat** [comment]
>Address items 1 and 5 from https://github.com/ClickHouse/ClickHouse/issues/50647.

@alexey-milovidov The metadata is still not presented, so it does not address 1

```
$ chl --path chl
create table data (key Int) engine=Log;
insert into data select * from numbers(100);
```

```
$ find chl
chl
chl/store
chl/store/9f5
chl/store/9f5/9f512eaf-db87-4d48-bd55-677d66a60621
chl/data
chl/data/system
chl/data/default
chl/data/default/data
chl/data/default/data/__marks.mrk
chl/data/default/data/sizes.json
chl/data/default/data/key.bin
chl/metadata
chl/metadata/information_schema.sql
chl/metadata/system
chl/metadata/INFORMATION_SCHEMA.sql
chl/metadata/system.sql
```

```
$ chl --path chl
ClickHouse local version 24.8.1.1.

:) show tables

SHOW TABLES

Query id: 7f010d67-fbeb-4b61-993d-231082b914c3

Ok.

0 rows in set. Elapsed: 0.002 sec.

:) select * from data

SELECT *
FROM data

Query id: 9050d1d5-1cbb-4820-b4ca-93b760c6c189


Elapsed: 0.053 sec.

Received exception:
Code: 60. DB::Exception: Unknown table expression identifier 'data' in scope SELECT * FROM data. (UNKNOWN_TABLE)

:) create table data (key Int) engine=Log

CREATE TABLE data
(
    `key` Int
)
ENGINE = Log

Query id: 27ea15d3-2fe3-4082-a044-54e9858cbe86


Elapsed: 0.018 sec.

Received exception:
Code: 57. DB::Exception: Directory for table data data/default/data/ already exists. (TABLE_ALREADY_EXISTS)
```

**Pull request merged**